### ISSUE TITLE
Coordinate system alignment - using 4 Marker Digitizing Probe

### ISSUE THREAD

**Issue by sciencecrazy** [volcic/motom-toolbox#12]
Hi Zoltan, 
Regarding defining a custom global coordinate system at the start of each test --> previously when I used the First Principles software, I would define an origin, and two of the axes, using a 4 marker digitizing probe (purchased directly from NDI). 
Could you please explain the process of defining a custom coordinate system using the API ? Do I need to create a new .ini file for the probe? [We have an Optotrak certus in the lab, and don't have the ndi digitising cube]

Thanks, 
Parham.

**Comment by sciencecrazy** [comment]
I use virtual markers (aka digitised points)a lot in my projects, so I would need to incorporate the digitising probe into my setup - I guess it would work in a similar way to the 6-marker VolcicLab Rigid Body which you showed in your paper?

**Comment by ha5dzs** [comment]
With the API, the coordinate system alignment is simply the process of assigning a rigid body's local coordinate system to the global coordinate system.

So, let's say you have an object and you attached at least 3 markers to it. This could be the Volciclab Universal Rigid Body, or some universal marker holder that you 3D printed. You don't have to get fancy expensive devices for this, as long as you can secure the markers so they won't move. I have a ton of these .stl files, if you want some, let me know.

To create the virtual marker, the easiest option is to create a local coordinate system for the rigid body, where the origin of the local coordinate system is where you want the rigid body's virtual marker to be. You can create this definition using NDI's 6D Architect software, which supports fancy transforms that are not implemented in the toolbox.

Then, once you have the rigid body and its definition file, you'll need to create a new .ini config file where you specify the number of markers, the marker/frame frequency, and the data collection time.

Afterwards, you can connect your markers to the strober, and call `optotrak_align_my_system()' with your rigid body file and .ini config file specified. When asked, position your rigid body where you want to have the global coordinate system's origin and axes to be, and press enter in the command window. This will start a small recording process (the length is specified in the .ini config file), and if all went well, it will create the new camera file that has your custom coordinate system alignment. If you have multiple cameras, you will be asked to do the registration process too, where you have to move the same rigid body around so both cameras can see it, and only then it will ask you to do the alignment.

You can use the newly created camera file in your optotrak setup.

In our lab, we use the NDI cube for this. The rigid body definition is such that the top end is the Z=0 plane, and the origin of its local coordinate system is one of the corners. We align its corner and its edges to the experimental table, and this creates our coordinate system definition that matches with the table.

I know that this process feels complicated, but once you have your object and its rigid body definition, you literally can create your new camera file in less than a minute. I do this every day when I have an experiment running.

**Comment by sciencecrazy** [comment]
Thanks for the detailed response Zoltan, appreciated. I was able to align the coordinate system to the local coordinate system of the rigid body as you mentioned. 
Just another semi related question here, regarding the config (.ini) file - in my experiments I use a lot of rigid bodies and LEDs and sometimes I have all my NDI smart markers attached to port 1 of the control unit and all the custom markers (3d printed rigid bodies with 3 LEDs attached to them) go into the strober and then to port 2 of the control unit. So I'm wondering how to define this in the config file? some channels on the strober are sometimes un-used, because some of the LEDs we bought from NDI were already wired in pairs, so say I have two 3d printed rigid bodies/marker carriers in an experiment, each with 3 LEDs on them: for marker 1: LED1= ch.1, LED2= ch.2, LED3= ch.3; for marker 2: LED1: ch.5, LED2: ch.6, LED3: ch.7. So essentially channel 4 wouldn't be connected to anything. And this was fine in the First Principles software because I was able to tell it exactly which channels on the strober were associated with which LEDs of which rigid body. Is this possible using the VI? Or does it just count every single channel?

**Comment by ha5dzs** [comment]
Great!

A far as I can tell, Optotrak requests you to disconnect any markers that are not part of the coordinate system alignment. If I have a large number of markers and rigid bodies in my experiment, I then just remove the strober altogether, and connect the alignment object via an other strober.

If that is not possible/inconvenient, you could probably play with the strober port table. I have not done this, so you are entering unchartered territory.

For example, you can have your experimental markers connected to the SCU's Port 2, and the markers used for alignment in port 1.

You can specify the markers by changing the `[Strober Table]` section of optotrak.ini (located in `C:\ndigital\settings` on Windows), or you could change the number of markers on each port using [`OptotrakSetStroberPortTable()`](https://github.com/volcic/motom-toolbox/wiki/API-functions-in-Matlab#optotraksetstroberporttable).

So, for alignment, you disable the markers connected to port 2, and for your experiment, you disable the markers connected to port 1. Then, assuming that the total number of markers on each port match with the config file, the system shouldn't complain about missing markers.

Again, I have not tried this, so chances are it might not work. Please let me know! :D

**Comment by sciencecrazy** [comment]
Hi Zoltan, cheers for that. I managed to get it working. I created two different config files for the test (all test markers) and digitisation (test markers+probe), then before running 'optotrak_set_up_system(camera_file, config_file)', I used 
'fail = OptotrakSetStroberPortTable(6,3,0,0);'  for the test, and
'fail = OptotrakSetStroberPortTable(6,3,4,0);'  for the digitisation
to tell it how many markers are connected on each port.
This way I didn't have to change the strober table in the optotrak.ini text file. 

I'll close the thread, and cheers for the help!
